Sofar: Add description (see discussion #458)

mathieumd opened the pull request.

mathieumd commented:
There is a problem with `group: Alert` (where the same registers are used several times)

mathieumd commented:
Fixed

davidrapan commented:
Hi @mathieumd, can you also try to merge all Fault X into single entity similar to Alert?

mathieumd commented:
I added phases (`l:`) and MPPT (`mppt:`). I also added batteries (`pack:`), but please double-chek I applied them correctly (please double-check the rest also, actually! ^^).

I don't see how to merge `Fault [1-12]`. Each of them uses `lookup:` tables. If we merge them (which is what does [the current `Alert`](https://github.com/davidrapan/ha-solarman/blob/main/custom_components/solarman/inverter_definitions/sofar_g3hyd.yaml#L885) already), how to distinguish theses tables by each register?

By the way, what's the purpose of `group:`? They don't seem to be used on HA webUI.

davidrapan commented:
Hi,

> I also added batteries (pack:)

I don't think that `pack:` is suitable here. If you only have one battery use `pack: 2` only for Battery 2 entities so it ensures that default configuration always show at least one.

> I don't see how to merge `Fault [1-12]`

Of course they can be merged you just need to combine all the lookup tables and shift their lookup keys according to register positions. Leave it be then If you don't know exactly how. 😉

> By the way, what's the purpose of `group:`?

It's just for configuring `update_interval`s, modbus function codes, etc. in bulk for multiple entities and so on.

mathieumd commented:
> I don't think that pack: is suitable here. If you only have one battery use only pack: 2 for Battery 2 entities so it ensures that default configuration always show at least Battery 1 entities.

But if you have *no* battery, you don't want any displayed.

> Of course they can be merged you just need to combine all the lookup tables and shift their lookup keys according to register positions. Leave it be then If you don't know exactly how. 😉

Something like that?!

```yaml
      - name: "Alert"
        rule: 6
        icon: "mdi:alert"
        registers:
          [
            0x0405,
            0x0406,
            0x0407,
            0x0408,
            0x0409,
            0x040A,
            0x040B,
            0x040C,
            0x040D,
            0x040E,
            0x040F,
            0x0410,
          ]
        lookup:

          # name: "Fault 1"
          # registers: [0x0405]
          - key: 0
            value: "No error"
          - key: 1
            value: "ID01 Grid Over Voltage Protection"
          - key: 2
            value: "ID02 Grid Under Voltage Protection"
          - key: 4
            value: "ID03 Grid Over Frequency Protection"
          - key: 8
            value: "ID04 Grid Under Frequency Protection"
          - key: 10
            value: "LOOKUP" # off-grid
          - key: 14
            value: "ID03 Grid Over Frequency Protection" # also ID14 Grid voltage unbalance
          - key: 16
            value: "ID05 Leakage current fault"
          - key: 32
            value: "ID06 High penetration error"
          - key: 64
            value: "ID07 Low penetration error"
          - key: 128
            value: "ID08 Islanding error"
          - key: 256
            value: "ID09 Grid voltage transient value overvoltage 1"
          - key: 512
            value: "ID10 Grid voltage transient value overvoltage 2"
          - key: 1024
            value: "ID11 Grid line voltage error"
          - key: 2048
            value: "ID12 Inverter voltage error"
          - key: 4096
            value: "ID13 Anti-backflow overload"
          - key: 8192
            value: "ID14"
          - key: 16384
            value: "ID15"
          - key: 32768
            value: "ID16"

          # name: "Fault 2"
          # registers: [0x0406]
          - key: 65536
            value: "No error"
          - key: 131072
            value: "ID17 Grid current sampling error"
          - key: 262144
            value: "ID18 Grid current DC component sampling error (AC side)"
          - key: 524288
            value: "ID19 Grid voltage sampling error (DC side)"
          - key: 1048576
            value: "ID20 Grid voltage sampling error (AC side)"
          - key: 2097152
            value: "ID21 Leakage current sampling error (DC side)"
          - key: 4194304
            value: "ID22 Leakage current sampling error (AC side)"
          - key: 8388608
            value: "ID23 Load voltage DC component sampling error"
          - key: 16777216
            value: "ID24 DC input current sampling error"
          - key: 33554432
            value: "ID25 DC component sampling error of grid current (DC side)"
          - key: 67108864
            value: "ID26 DC input branch current sampling error"
          - key: 134217728
            value: "ID27"
          - key: 268435456
            value: "ID28"
          - key: 536870912
            value: "ID29 Leakage current consistency error"
          - key: 1073741824
            value: "ID30 Grid voltage consistency error"
          - key: 2147483648
            value: "ID31 DCI consistency error"
          - key: 4294967296
            value: "ID32"
```

davidrapan commented:
Yes! That looks good. But I just noticed... you need to make the PR from some other branch than `main`.

mathieumd commented:
Then continuing to https://github.com/davidrapan/ha-solarman/pull/494
